### ISSUE TITLE
Feature: Disk write protect (like switch on real floppy disks)

### ISSUE THREAD

**Issue by EMLommers** [openMSX/openMSX#1514]
When selecting DirAsDisk, have the possibility to 'mount' the directory as R/O.
This to ensure nothing happens to this directory and sub-directories.

Even the xml of a certain machine mentioned amount of drives is 1, the possibility to override this variable and 'mount' another DirAsDisk to the B: drive.

With this possibility you ensure you can read any file from A: , with R/O and can save file to the DirAsDisk selected as B: drive.

**Comment by m9710797** [comment]
Hi,

There are two separate questions in this ticket:

1) Mount DirAsDsk read-only:
That's already possible via the setting `set DirAsDSKmode read_only`. Note that it only makes the disk read-only from the MSX side. Files can still be changed on the host side, and those changes will be reflected in the virtual disk. See here for a more detailed discussion (last 4 comments):
    https://github.com/openMSX/openMSX/issues/1410

2) I'm not sure I've correctly understood your second question. You want to use the MSX machine as-if it has two drives even though in the machine configuration file it's stated that there's only 1 drive, correct?  Possibly only for DirAsDsk?? I'll try to answer, but my apologies if I misunderstood the request.

Changing the numbers of drives in a running MSX machine is difficult. It requires changing various internal BDOS state. The details also differ a lot between different versions (DOS1, DOS2, Nextor, ...). This will be tricky to get right. And if the goal is to make using DirAsDsk safer, then this approach will possibly do more harm than good.

So let's instead assume you meant overriding the number of drives when the MSX machine starts (not when it's already running). But then there already is an option to do that: just edit the machine config file. (Either change the file directly, or make a copy and edit that).

I do understand your motivation for this request. You lost a lot of your (host) files by using DirAsDsk(*1) (I'm very sorry for that) and you want to make it safer to use DirAsDsk. I do support your goal, but I don't believe this (2nd) request is the right approach.

(*1) The host file corruption you got was caused by a bug in DirAsDsk. That bug has been fixed in openMSX version 18.
We do our best, but we can't 100% guarantee that it's now fully bug free, or that we'll never accidentally introduce new bugs in the future. But to mitigate the harm that such bugs can do, we're currently thinking about this solution:
   https://github.com/openMSX/openMSX/issues/1512
In summary: we'll no longer import partial host files in the virtual DirAsDsk (e.g. when the disk is full). That way we'll never write out such truncated files back to the host.

Wouter

**Comment by EMLommers** [comment]
Hi Wouter, 
Yes I have tested 18.0 and it was resolved! 
Also no files were truncated... Glad!  Prior to 18.0 all files larger than 1.0kb were truncated recursively to 1.0kb when drive was full..  
Maybe it is a Catapult feature, did not know where to put it. 

Clear. I will make separate ticket for:
Override xml in number of drives when selected b: drive to 2 if machine has only 1.
I tested with Sony HB-F700D changed amount of drives to 2 and had A and B drive. 
Maybe it depends on the diskrom, whether it is possible or not. 

Regards
Ernst

**Comment by m9710797** [comment]
Yes, this absolutely depends on the disk rom.
I'm half guessing here, but if I, as a MSX machine manufacturer, don't ship any MSX models with two disk drives, then why should I then spend time writing software (the disk rom) for supporting two drives. So I can imagine that this indeed won't work an all MSX machines.

In general, in openMSX we (try to) emulate the MSX hardware, but we (try to) make no assumptions about the software that is running on the machine. That way openMSX should work equally well with the standard MSX software (DOS1, DOS2) or with 'future' software like Fuzix, or even totally different operating systems like SymbOS or Uzix. The latter supports disk drives in a totally different way (there is no drive 'A' or 'B'). So even if we wanted to, it's simply impossible for openMSX to override the software interpretation(*1) of the number of drives.
(*1) Thus changing the interpretation after the machine has booted.

About overriding features in the machine description file. Maybe I'm still not understanding your request. But the way I understand it now, I don't think it's a good idea. For example a standard Philips_NMS-8250 has a 128kB memory mapper. We _could_ make an option to override that to have a larger memory mapper. But then it's no longer a standard 8250 machine. You can instead make a new machine config file with a different value for the memory size. It's the equivalent of soldering a larger memory mapper in a real 8250 machine, but that's then a modified 8250.

Similar for the number of drives in the config file. We _could_ make an option for that. But why? It's a lot of work for a very specific thing (very few people need this).  Especially because there already is a very easy workaround: just make a copy of the machine config and edit that.

**Comment by EMLommers** [comment]
Yes, see my issue Sony HB-F700D Rom.
I added a xml file like you mentioned.
That is the reason also about Gui Machine Config.
You can emulate as close as possible with 1001 different roms, or just combine sometimes the best config you desire, but still MSXx.
So select engine, diskrom, extrom, drives etc. MSX bios is quite standard. 
Add basic kun rom, fmpac etc. 
This enables people also to select 'standard' extensions and 'non standard' extensions. 
If you can add memory in xml, why add extension of 512Kb or different memory size. 
Same for fmpac, sram, Vram, etc.... 
Why make it users difficult by selecting 1001 extensions.... 
Of course I understand you like to add all the cartridge and other external extensions made by all manufacturers. Isn't it overkill... 

Also if you want to emulate everything you should also be able to select for example Fast DiskROM, as some have changed

Regards 
Ernst

**Comment by m9710797** [comment]
OK, I see now.

I understand your request. You want to use "super" or "boosted" MSX machines. It's a valid approach, but it's not the way we usually handle things in openMSX. Our approach is to stay as close as possible to real/existing machines/extensions as possible (or at least we try to).  (*1)

That said, we do also offer some "Boosted" machine configurations (they offer as many hardware features as possible, but still remain compatible with as much MSX software as possible, not always an easy balance). Maybe one of these machines already suits your needs?

And you're right: it would be nice to have a Gui machine config editor. But it's not something I personally plan to work on. (Help is always welcome of course).



(*1) Many of the bug reports we receive also happen on the corresponding real MSX machine. It would be much harder to verify such bugs if we can't easily test on the same real and emulated hardware.

**Comment by EMLommers** [comment]
On the other hand with a Machine Config Editor, you can test much quicker several machines, just selecting different VDP, ram, vram and run your routine again.
Easier than selecting Panasonic, Sony, Sanyo (MSX1, 2,2+,tr)

I was planning to create this, but was informed OpenMSX was moving away from wx to another framework, I like to keep inline and not do work twice due to switching framework. Qt?

It is easier than editing a xml like attached.

[Sony_HB-F700D_2+.xml.txt](https://github.com/openMSX/openMSX/files/11856360/Sony_HB-F700D_2%2B.xml.txt)

Xml for converted Sony to 2+.  Lazy to extract roms from real machine so used cross manuf. Roms and roms from bluemsx

**Comment by MBilderbeek** [comment]
If I understand correctly, can we close this ticket and perhaps open a new one with a feature request to create an openMSX hardware configuration editor?

**Comment by EMLommers** [comment]
This should feature should be opened with Catapult. It is in Catapult not possible to open drive as R/O
I have already made ticket for GUI config editor

**Comment by MBilderbeek** [comment]
If it is Catapult specific, then please make an issue in the Catapult issue tracker. However, please note that the Catapult project is in maintenance mode only. No features will be added, at least not by any active developer I know.

**Comment by MBilderbeek** [comment]
By the way, which was the issue about the GUI config editor?

**Comment by EMLommers** [comment]
Which framework will you shift to?
So I can start making the config editor... wx,qt,other?

**Comment by MBilderbeek** [comment]
Do you want to write an external program, independent of openMSX? Or make it part of the internal GUI?

In the former case, you can use any framework you like. In the latter case, we're moving towards DearImGUI, you can check it out on its own branch.

**Comment by m9710797** [comment]
We're indeed doing an experiment with "Dear ImGui". You can find it in the "imgui" branch in the openMSX git repo. Feel free to check it out, and give feedback.

The experiment is going very well. +90% of the functionality of the current catapult and of the current debugger application is already ported. But it still needs a lot of fine-tuning. (But subjectively, IMHO, some parts, especially in the debugger, already work better than in the current applications).

**Comment by EMLommers** [comment]
I was thinking to modify the current setup a bit.
We are talking about Catapult now , in my opinion.

At the tab Session, I was planning to modify it a bit and and make sub-tabs.
So you can create several configurations (I would like it to be sessions, with different configurations.)

Similar like aqemu or virtualbox.
Selections of System ROMs , add ROMs, memory, etc.

What I am missing in openMSX is like virtualbox, when closing the window, dialog if you really want to quit, save state or exit.
I have had several times, that I accidentally closed the window ( X ) instead of lowering  ( _ )..... and current state is gone..

**Comment by EMLommers** [comment]
Am I correct the new version of openMSX has catapult embedded?

**Comment by m9710797** [comment]
Correct. (And then obviously it's then no longer called 'catapult', it's just openMSX).
Same for the debugger.

(3rdparty) external applications that control openMSX will continue to be supported. But the default GUI ("catapult" as you call it) won't use it anymore.

**Comment by EMLommers** [comment]
I have build the ImGui version, took some time on this cheap laptop, and indeed it is embedded.
The feeling, looks and user friendliness  is still far from wx. of course this is the beginning.
Why the change?
Catapult was more, how to call it, inviting to use. 
I guess layout, looks and feel will be upgraded in the future. I really have to get used to it.

I will start reading the code and start building config editor.

You have now included save state, etc (Snapshot) , Really good!
Wow! The functions are really extended compared to Catapult.

And multiple instances!! Good job!
Tabs are great!

The first time I started the new version it could not find the ROM files, but the second time i started ROMs were found.

overall, I have to get used to it, but looks very promising! 

Is the OSD also changing or stays as it is?

**Comment by MBilderbeek** [comment]
I think it would be helpful not to mix things up and name things by their name.
- Catapult is an external GUI application to control openMSX, which started as a launcher. It's written in wxWidgets but shows a native look and feel. There hasn't been made any development on it for many years.
- There is the built-in OSD menu, written in Tcl. It's quite a limited environment, using only very basic graphical objects. Started as an experiment, but it's already quite extensive and has functionality not found in Catapult (e.g. save states)
- There are some other things like tools and gimmicks written in Tcl with OSD elements. Including a tabbed browsing thing.... (as @EMLommers calls "Multiple instances")
- And there's the new DearImGUI user interface. It's written in C++ using a totally different paradigm and works only with OpenGL rendering. An experimental branch is available, but it's far from ready to merge. In the future, we will integrate it and call it "the openMSX GUI". Catapult is then obsoleted and probably also the separate debugger GUI. And also the OSD menu, which is removed in that branch. Still, lots of useful tool scripts will remain and they use the OSD mechanisms.

I was very confused you started to talk about Catapult but actually meant the OSD menu. 

Why the change? Because of being directly integrated into openMSX and the immediate mode GUI paradigm, it is much much easier to accomplish more complex things with much better performance (e.g. real time monitoring/visualization of data).

**Comment by EMLommers** [comment]
In the beginning I thought the OSD menu was part of Catapult. As a layer over the running instance with menu button.
OSD is not part of Catapult I understand now..
As I like to use GUI when available. Was not aware OSD was separated from Catapult. as the same options were available.

 Will check the code.
Looks very promising!

**Comment by EMLommers** [comment]
Using again SDL2 or also OpenGL3?

**Comment by m9710797** [comment]
Both SDL2 and OpenGL-2.x.

OpenGL is a "graphics backend". It's very good at rendering, but can only do rendering.
SDL2 is a "platform backend". It handles stuff like window creation, keyboard/mouse input, ... (And yes, SDL2 could also handle some simple graphics, but not (high performace) enough for e.g. Dear ImGui).

"Dear ImGui" supports various different backends: https://github.com/ocornut/imgui/blob/master/docs/BACKENDS.md
And in case of openMSX we've choosen for the combination SDL/OpenGL because that's cross-platform.

**Comment by EMLommers** [comment]
![Screenshot from 2023-06-26 08-41-17](https://github.com/openMSX/openMSX/assets/79648274/4392ea14-19ab-4cad-b2bd-deecbf575286)
As I was bit skeptic about the style of the windows, and noticed a lot of issues with style, the ugly triangles, but everything is customizable I found out. Playing with the demo setup and modifying code.
As you can see in the snapshot style is so damn important to me, That was the first thing I have been researching on, style.. The only thing I had not changed was the font. When changing font and add shadow to window, it looks superb !
The possibilities are huge, and I think a very good choice (now I know the style of windows is modify able, and a lot of other things as welll.)  
So I am already doing some homework. And by the way chatGPT is quite handy as co-worker ;)
The examples he gives is primary SDL2 and OpenGl3, if you explicitly ask for SDL2 and OpenGL2, he provides the code.
For getting familiar it is a really nice companion 

Are you using tinyxml2 for xml parsing or a different one? Or does it not matter which one I will use.
[https://github.com/leethomason/tinyxml2](url)

As for the config xml's I noticed you are not using closing tags all the time. Is there a reason for it? 

First I will check the code and work on a style I prefer, and create the config editor, at a later stage you can change it like your preferences. Will you include a skin option? To change all styles?
